server/table_namespace_classifier.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"github.com/pingcap/pd/server/core"
+	"github.com/pingcap/pd/server/namespace"
+)
+
+type tableNamespaceClassifier struct {
+	nsInfo         *namespacesInfo
+	tableIDDecoder core.TableIDDecoder
+}
+
+func newTableNamespaceClassifier(nsInfo *namespacesInfo, tableIDDecoder core.TableIDDecoder) tableNamespaceClassifier {
+	return tableNamespaceClassifier{
+		nsInfo,
+		tableIDDecoder,
+	}
+}
+
+func (c tableNamespaceClassifier) GetAllNamespaces() []string {
+	nsList := []string{}
+	for name := range c.nsInfo.namespaces {
+		nsList = append(nsList, name)
+	}
+	return nsList
+}
+
+func (c tableNamespaceClassifier) GetStoreNamespace(storeInfo *core.StoreInfo) string {
+	for name, ns := range c.nsInfo.namespaces {
+		for _, storeID := range ns.StoreIDs {
+			if storeID == storeInfo.Id {
+				return name
+			}
+		}
+	}
+	return namespace.DefaultNamespace
+}
+
+func (c tableNamespaceClassifier) GetRegionNamespace(regionInfo *core.RegionInfo) string {
+	for name, ns := range c.nsInfo.namespaces {
+		startTable := c.tableIDDecoder.DecodeTableID(regionInfo.StartKey)
+		endTable := c.tableIDDecoder.DecodeTableID(regionInfo.EndKey)
```

Comment:
We can move the 2 lines out of for loop. Also we can check if start == end before the loop, if they are not equal, we can return defaultNamespace directly.